### ISSUE TITLE
Please consider to not deprecate pre-link

### ISSUE THREAD

**Issue by johanneskoester** [conda/conda#7452]
**I'm submitting a...**
  - [ ] bug report
  - [x] feature request

## Current Behavior
Conda currently prints a [warning](https://github.com/conda/conda/blob/0ee04d96fd761753085dfad3f8644f0edb997b16/conda/core/link.py#L770) when a pre-link script is used in a package.
The warning mentions that pre-link might be deprecated in the future.

## Desired Behavior
The [GGD project](https://github.com/gogetdata/ggd-recipes), a project that complements Bioconda with packages that provide scientific data (e.g. reference genome files), makes heavy use of pre-link scripts.
They are needed to download datasets from persistent sources upon installation of a GGD Conda package. This keeps the Conda-Package itself lightweight. While delivering the same experience as if the dataset would be part of the package. Since these datasets can be huge, up to tens of gigabytes, it can be very beneficial to avoid redundant data storage.

Hence, we would appreciate a lot if pre-link would not be deprecated and the warning would be removed. Alternatively, maybe a restricted version of pre-link that does not have the current issues could be considered?

**Comment by johanneskoester** [comment]
@kalefranz could you have a look at this? Thank you!

**Comment by kalefranz** [comment]
@johanneskoester I've looked briefly at a couple of the examples you link to in the GGD project, and it's not immediately clear to me why there has been a choice of pre-link over post-link.  Or, alternately, why post-link is not equivalent to pre-link in this case.  Note that post-link is not being deprecated, only pre-link.

The reason behind deprecating pre-link is that it is primarily used to make changes to the package cache rather than the install prefix, often in non-idempotent ways, and can therefore have an effect on *all* conda environments a package is installed into, not just the target environment of the current conda invocation.

What's being done here may well end up being a legitimate use case for pre-link, and I'm definitely open to finding an alternate solution than blanket deprecation if that turns out to be true.

**Comment by kalefranz** [comment]
Looking specifically at https://github.com/gogetdata/ggd-recipes/tree/master/recipes/Canis_familiaris/canFam3/canfam3-dbsnp for example, it looks to me like what you actually *want* is post-link rather than pre-link.  The meta.yaml lists `gsort` as a run requirement, and I see `gsort` being used in the script, but that's only guaranteed to be available if the script is post-link and not pre-link.

**Comment by kalefranz** [comment]
I see some other details that might also go against what I'd consider best practices for post-link scripts.  We should really right up something for a docs about what we recommend as best practices for post-link, how to use it correctly, and different things we've seen that we consider anti-patterns.  I'll add that to the long TODO list :)

**Comment by kalefranz** [comment]
Might as well make some notes for that documentation here I guess.  With regard to: https://github.com/gogetdata/ggd-recipes/blob/38822e9a4963b54fc359be7a3b04c110d53b6bd9/recipes/Canis_familiaris/canFam3/canfam3-dbsnp/pre-link.sh

* Don't target for bash. Target for generic/posix /bin/sh.  Make use of https://www.shellcheck.net/ with `#!/bin/sh` as the shebang.
* Make sure the scope of the script, any side effects, are confined to the target $PREFIX. Conda post-link shouldn't be used as configuration management, and post-link shouldn't have side effects outside of the target environment or system-wide. I don't think this specific example has an issue with that.
* Any commands invoked within the script that are not shell built-ins should be provided as run requirements. This script has `gsort`, for example, but makes the assumption that `wget` is universally available, which it really shouldn't.

**Comment by johanneskoester** [comment]
Thanks a lot for the feedback, this is greatly appreciated!
The reason to choose pre-link is that these scripts can be expensive sometimes. What GGD is doing:

The recipes describe e.g. reference genomes, which are hosted at institutions like the NIH. Such a recipe shall allow people to easily obtain a copy of the reference genome. In order to safe space, the reference genome (which can be several GB) is not included in the package, but rather the package just delivers the code to download the genome from the NIH. This download (and eventual format conversions) happens on the fly in the pre-link script. Since it can take some time, we chose pre-link, such that different conda environments and users can benefit from the download already been done on the first installation. At least that was our idea.

Does that make sense, or do I have a misconception about pre-link?

**Comment by mike-wendt** [comment]
+1

Would be helpful for NVIDIA as well to post a EULA message prior to install for interactive users and to print the output during `-y` installs.

**Comment by jakirkham** [comment]
Yeah IDK if this change is still being considered (the post is from 2018 and the warning from 2016). If so, it is worth noting some conda-forge recipes make use of this functionality too.

**Comment by marcelotrevisani** [comment]
I believe it would be interesting to add a new key in the recipe, to be specific in the ``about`` section, something like ``eula`` which points to a txt file.
so when we create a package we can add that msg inside of the ``eula`` file to be shown when installing the package.
I think for this implementation we will need to modify ``conda`` and ``conda-build``

**Comment by jakirkham** [comment]
(bumping to keep alive)

**Comment by marcelotrevisani** [comment]
I think this can help as well
https://github.com/conda/conda/pull/11123

**Comment by marcelotrevisani** [comment]
after we integrate #11123 can we close this issue?

**Comment by jakirkham** [comment]
Does this require changes in Conda to work as well?

**Comment by marcelotrevisani** [comment]
> Does this require changes in Conda to work as well?

What do you mean?

**Comment by jakirkham** [comment]
Ah sorry nvm
